README
Sinatra webapp for a file-based CMS

data/history.txt
1993 - Yukihiro Matsumoto dreams up Ruby.
1995 - Ruby 0.95 released.
1996 - Ruby 1.0 released.asdasd
1998 - Ruby 1.2 released.
1999 - Ruby 1.4 released.
2000 - Ruby 1.6 released.
2003 - Ruby 1.8 released.
2007 - Ruby 1.9 released.
2013 - Ruby 2.0 released.
2013 - Ruby 2.1 released.
2014 - Ruby 2.2 released.
2015 - Ruby 2.3 released.
2016 - Ruby 2.4 released.
2017 - Ruby 2.5 released.
2018 - Ruby 2.6 released.
2019 - Ruby 2.7 released.
2020 - Ruby 3.0 released.
2021 - Ruby 3.1 released.
2022 - Ruby 3.2 released.

test/app_test.rb
ENV["RACK_ENV"] = "test"

require "minitest/autorun"
require "minitest/reporters"
require "rack/test"
require "fileutils"

require "pry"
require "pry-byebug"

Minitest::Reporters.use!

require_relative "../app.rb"

class AppTest < Minitest::Test
  include Rack::Test::Methods

  def session
    last_request.env["rack.session"]
  end

  def create_document(name, content = "")
    File.open(File.join(data_path, name), "w") do |file|
      file.write(content)
    end
  end

  def admin_session
    { "rack.session" => { username: "admin" } }
  end

  def app
    Sinatra::Application
  end

  def setup
    FileUtils.mkdir_p(data_path)

    create_document "about.md", "# ruby is"
    create_document "history.txt", "1993 - Yukihiro Matsumoto dreams up Ruby.\n 1995 - Ruby 0.95 released.\n 1996 - Ruby 1.0 released.\n "
    create_document "changes.txt", "2020 - Ruby 3.0 released.\n 2021 - Ruby 3.1 released.\n 2022 - Ruby 3.2 released.\n "
    create_document "xyz.xyz", "aaaaa"

    pattern = File.join(data_path, "*")

    @files = Dir.glob(pattern).map do |path|
        File.basename(path)
    end
  end

  def teardown
    FileUtils.rm_rf(data_path)
  end


  def test_index
    get "/", {}, admin_session

    assert_equal 200, last_response.status
    assert_equal "text/html;charset=utf-8", last_response["Content-Type"]
    assert_includes last_response.body, "about.md"
    assert_includes last_response.body, "changes.txt"
  end

  #test non-existent files
  def test_invalid_file
    get "/notgonnahappen132323.txt", {}, admin_session

    assert_equal 302, last_response.status
    assert_equal session[:error], "notgonnahappen132323.txt does not exist!"
  end

  #test if the markdown files are display as HTML
  def test_md_file
    get "/about.md", {}, admin_session

    assert_equal 200, last_response.status
    refute_includes last_response.body, "# ruby is  "
    assert_includes last_response.body, "<h1>ruby is</h1>"
  end

  def test_unkown_extension
    get "/xyz.xyz", {}, admin_session

    assert_equal session[:error], "Sorry, the .xyz file format is not supported yet, check back again soon!"
  end


  #test the edit page
  def test_edit_page
    @files.each do |file_name|
      next if file_name == "xyz.xyz" #since file is not supported

      file_path = File.join(data_path, file_name)
      file_content = File.read(file_path)

      get "/#{file_name}/edit", {}, admin_session
      extension = File.extname(file_name)[1..-1]

      assert_equal 200, last_response.status
      assert_includes last_response.body, "Editing content of #{file_name}"
      assert_includes last_response.body, file_content
      assert_includes last_response.body, "<textarea name=\"content\" rows=\"5\" cols=\"60\" class=\"#{extension}\">#{file_content}</textarea>"
      assert_includes last_response.body, "<form action=\"/#{file_name}/edit\""
    end
  end

#test if the post method works and changes the contents of file
  def test_edit_file
    @files.each do |file_name|
      next if file_name == "xyz.xyz" #since file is not supported

      file_path = File.join(data_path, file_name)
      file_content = File.read(file_path)
      new_content = "The #{file_name} has been tampered with!" + file_content

      post "/#{file_name}/edit", {content: new_content}, admin_session
      assert_equal 302, last_response.status
      assert_equal "The #{file_name} has been updated!", session[:success]

      get "/#{file_name}"

      assert_equal 200, last_response.status
      assert_includes last_response.body, "The #{file_name} has been tampered with!"

      get "/#{file_name}/edit"
      extension = File.extname(file_name)[1..-1]

      assert_equal 200, last_response.status
      assert_includes last_response.body, "<textarea name=\"content\" rows=\"5\" cols=\"60\" class=\"#{extension}\">#{new_content}</textarea>"
      assert_includes last_response.body, "<form action=\"/#{file_name}/edit\""
    end
  end

  #test new document creation
  def test_create_new_file
    post "/new", {file_name: "test.txt"}, admin_session

    assert_equal 302, last_response.status
    assert_equal "test.txt has been created!", session[:success]

    get last_response["Location"]

    assert_includes last_response.body, "test.txt"
    assert_equal 200, last_response.status

    get "/test.txt"
    assert_equal 200, last_response.status
  end

  # test new document creation without a valid extension
  def test_creation_valid_extension
    post "/new", {file_name: "arbitarysomething"}, admin_session

    assert_equal 422, last_response.status
    assert_includes last_response.body, "Sorry that extension is not supported yet!"
    # assert_equal "Sorry that extension is not supported yet!", session[:error]
    # The above line can not be used because session[:error] is already deleted to display the error.

    get "/"
    refute_includes last_response.body, "arbitarysomething"
  end

  #test empty string file_name

  def test_creation_empty_file_name
    post "/new", {file_name: ""}, admin_session

    assert_equal 422, last_response.status
    assert_includes last_response.body, "Sorry, you must enter a file name"
  end

  # checks if files have been deleted
  def test_deletion
    @files.each do |file_name|
      post "/#{file_name}/delete", {delete: true}, admin_session

      assert_equal 302, last_response.status
      assert_equal "The #{file_name} document has been deleted!", session[:success]

      get last_response["Location"], {}, admin_session

      refute_includes last_response.body, "/#{file_name}/edit"
    end
  end

  # test if the signin page is displaying correctly
  def test_login_page
    get "/users/login"

    assert_equal 200, last_response.status
    assert_includes last_response.body, "<input type=\"text\" name=\"password\">"
    assert_includes last_response.body, "<input type=\"text\" name=\"username\" value=\""
  end

  # test if the user is notified about invalid credentials and can retry
  def test_login_invalid
    post "/users/login", username: "admin", password: "test"

    assert_equal 401, last_response.status
    assert_includes last_response.body, "Invalid credentials"

    assert_includes last_response.body, "<input type=\"text\" name=\"password\">"
    assert_includes last_response.body, "<input type=\"text\" name=\"username\" value=\"admin\""
    refute session[:loggedin]
  end

  # test if the user can sign in with the correct credentials
  def test_login_valid
    post "/users/login", username: "admin", password: "secret"

    assert_equal 302, last_response.status
    assert_equal "Successfully logged in as admin", session[:success]
    assert_equal "admin", session[:username]

    get last_response["Location"]

    assert_equal 200, last_response.status
    @files.each do |file_name|
      assert_includes last_response.body, "<a href=\"/#{file_name}/edit\">Edit</a>"
   end

    assert_includes last_response.body, "Signed in as admin"
  end

  # test if the user can signout
  def test_signout
    post "/users/logout", {}, admin_session
    assert_equal "You've been logged out", session[:success]
    assert_nil session[:username]

    get last_response["Location"]
  end

  def test_unauth_index
    get "/"
    assert_equal 302, last_response.status
    assert_equal "You must be signed in to do that.", session[:error]

    get last_response["Location"]

    refute session[:loggedin]
    refute session[:username]
    assert_includes last_response.body, "<input type=\"text\" name=\"password\">"
    assert_includes last_response.body, "<input type=\"text\" name=\"username\""
  end

  # testin unautharized form submission and page viewing

  def test_unauth_new_file_page
    get "/new"
    assert_equal 302, last_response.status
    assert_equal "You must be signed in to do that.", session[:error]
  end

  def test_aunauth_new_file
    post "/new", file_name: "testingtestingTESTING.txt"
    assert_equal 302, last_response.status
    assert_equal "You must be signed in to do that.", session[:error]

    get "/", {}, admin_session
    refute_includes last_response.body, "testingtestingTESTING.txt"
  end

  def test_unauth_edit_page
    get "/about.txt/edit"
    assert_equal 302, last_response.status
    assert_equal "You must be signed in to do that.", session[:error]
  end

  def test_unauth_edit
    @files.each do |file_name|
      post "/#{file_name}/edit", content: "testsabcdtestefghtest"
      assert_equal 302, last_response.status
      assert_equal "You must be signed in to do that.", session[:error]
    end

    get "/", {}, admin_session

    @files.each do |file_name|
      get "/#{file_name}"
      refute_includes last_response.body, "testsabcdtestefghtest"
    end
  end

  def test_unauth_delete
    @files.each do |file_name|
      post "/#{file_name}/delete"
      assert_equal 302, last_response.status
      assert_equal "You must be signed in to do that.", session[:error]
    end
  end
end


app.rb
require "sinatra"
require "sinatra/reloader"
require "sinatra/content_for"
require "tilt/erubis"
require "redcarpet"
require "psych"

def data_path
  if ENV["RACK_ENV"] == "test"
    File.expand_path("../test/data", __FILE__)
  else
    File.expand_path("../data", __FILE__)
  end
end

configure do
  enable :sessions
  set :session_secret, "secret"
end

before do
  pattern = File.join(data_path, "*")

  @files = Dir.glob(pattern).map do |path|
      File.basename(path)
  end
end

helpers do
  def signed_in?
    !!session[:username]
  end
end


# find extension of file and returns its contents formatted
def load_file_content(file_path)
  extension = File.extname(file_path)

  case extension
  when ".txt"
    response.headers["Content-Type"] = "text/plain"
    File.read(file_path)
  when ".md"
    markdown = Redcarpet::Markdown.new(Redcarpet::Render::HTML)
    markdown.render(File.read(file_path))
  else
    session[:error] = "Sorry, the #{extension} file format is not supported yet, check back again soon!"
    redirect "/"
  end
end

# load_file_content but raw data
def raw_load_file_content(file_path)
  extension = File.extname(file_path)

  case extension
  when ".txt", ".md"
    File.read(file_path)
  else
    session[:error] = "Sorry, the #{extension} file format is not supported yet, check back again soon!"
    redirect "/"
  end
end

# Checks if file exists, creates error mesage if file does not exist.
def file_exists?(file_path)
  File.file?(file_path)
end

def valid_authentication(username, password)
  return false if username.strip == ""
  Psych.load_file("../data/users.yml")[username] == password
end

# new document form
get "/new" do
  redirect_logged_out unless signed_in?
  erb :new
end

#create a document with name argument
def create_document(file_name, content = "")
  File.open(File.join(data_path, file_name), "w") do |file|
    file.write(content)
  end
end

def valid_extension?(file_name)
  file_path = File.extname(file_name)
  file_path == ".md" || file_path == ".txt"
end

# create new document
post "/new" do
  redirect_logged_out unless signed_in?

  file_name = params[:file_name]
  file_path = File.join(data_path, file_name)

  if file_exists?(file_path)
    session[:error] = "#{file_name} already exist!"
  elsif file_name == ""
    session[:error] = "Sorry, you must enter a file name"
  elsif !valid_extension?(file_name)
    session[:error] = "Sorry that extension is not supported yet!"
  else
    session[:success] = "#{file_name} has been created!"
    create_document(file_name)
    status 302
    redirect "/"
  end

  status 422
  erb :new
end

def redirect_logged_out
  session[:error] = "You must be signed in to do that."
  redirect "/users/login"
end

# load index and show files available in CMS
get "/" do
  redirect_logged_out unless signed_in?
  erb :index
end

#load file depending on the extension
get "/:filename" do |filename|
  redirect_logged_out unless signed_in?

  @file_path = File.join(data_path, filename)

  unless file_exists?(@file_path)
    session[:error] = "#{filename} does not exist!"
    redirect "/"
  end

  load_file_content(@file_path)
end

#display file contents to be edited
get "/:filename/edit" do
  redirect_logged_out unless signed_in?

  @filename = params[:filename]
  file_path = File.join(data_path, @filename)

  @extension = File.extname(@filename)[1..-1]

  unless file_exists?(file_path)
    session[:error] = "#{filename} does not exist!"
    redirect "/"
  end

  @file_content = raw_load_file_content(file_path)
  session[:success] = "The #{@filename} has been edited!"

  erb :edit
end

#update file contents
post "/:filename/edit" do
  redirect_logged_out unless signed_in?

  @filename = params[:filename]
  file_path = File.join(data_path, @filename)

  file_content = params[:content]

  File.write(file_path, file_content, mode: "w+")
  session[:success] = "The #{@filename} has been updated!"

  redirect "/"
end

# delete file
post "/:filename/delete" do
  redirect_logged_out unless signed_in?

  @filename = params[:filename]
  file_path = File.join(data_path, @filename)

  if params[:delete]
    session[:success] = "The #{@filename} document has been deleted!"
    File.delete(file_path)
  end

  redirect "/"
end

# display login page
get "/users/login" do
  erb :login
end

# login the user
post "/users/login" do
  @username = params[:username]
  password = params[:password]

  if valid_authentication(@username, password)
    session[:success] = "Successfully logged in as #{@username}"
    session[:username] = @username
    redirect "/"
  else
    session[:error] = "Invalid credentials"
    status 401
    erb :login
  end
end

post "/users/logout" do
  if session[:username]
    session[:username] = nil

    session[:success] = "You've been logged out"
  end

  redirect "/users/login"
end